src/js/common/schemaform/FileField.jsx

```diff
@@ -0,0 +1,161 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import _ from 'lodash/fp';
+// import Scroll from 'react-scroll';
+import { displayFileSize } from '../../common/utils/helpers';
+import environment from '../../common/helpers/environment';
+
+export default class FileField extends React.Component {
+  onAddFile = (e) => {
+    const files = this.props.formData || [];
+    this.props.onChange(files.concat({
+      uploading: true
+    }));
+    this.uploadFile(e.target.files[0])
+      .then(fileInfo => {
+        const uploadingFiles = this.props.formData;
+        this.props.onChange(_.set(uploadingFiles.length - 1, {
+          errorMessage: null,
+          uploading: false,
+          fileName: fileInfo.fileName,
+          maxSize: fileInfo.maxSize,
+          confirmationNumber: fileInfo.confirmationNumber
+        }, uploadingFiles));
+      })
+      .catch(message => {
+        const uploadingFiles = this.props.formData || [];
```

Comment:
Why define `uploadingFiles` here (and in the `.then()` above) instead of using `files`? I mean, we're using lodash/fp, so the `_.set()` on it should return a new array so immutability isn't the issue, yeah? (And if it was, I suppose we'd need to clone `this.props.formData` anyhow, eh?)

My guess is it's because of the possibility of having different props by the time the promise resolves or rejects. Is that it?